components/feature/push/src/main/java/mozilla/components/feature/push/AutoPushFeature.kt

```diff
@@ -359,7 +361,21 @@ internal fun CoroutineScope.launchAndTry(
         try {
             block()
         } catch (e: RustPushError) {
-            if (e !is CommunicationServerError && e !is CommunicationError && e !is RecordNotFoundError) {
+            val result = when (e) {
+                is CryptoError,
+                is CommunicationError,
+                is CommunicationServerError,
+                is AlreadyRegisteredError,
+                is StorageError,
+                is MissingRegistrationTokenError,
+                is StorageSqlError,
+                is TranscodingError,
+                is RecordNotFoundError,
+                is UrlParseError -> false
+                else -> true
```

Comment:
So, new exception types will throw? Not sure that's worth creating us extra work in the future (to keep this list up-to-date). Either way, these exceptions will end-up in Sentry - either as `info`s, or as `fatal`.

There's an idea in building robust software systems - be lenient with what you accept, and strict with what you send out. I think it applies here. We're handling an external message, subject to all sorts of unexpected issues outside of our control. If we crash here in response to a bad message, we're not gaining much! So, maybe our push stack got into a pickle and doesn't function correctly (either due to a server issue, or a client issue, or some network corruption, or..) - what do we gain by crashing here? The rest of the app is, most likely, perfectly functional. You'll just be getting in a user's way.

As long as we're aware of these problems, we can hopefully fix them - and that's what `submitCaughtException` already buys you. Submit these would-be crashes as `info`s, and let the app live :)

Reply:
From [the original discussion](https://github.com/mozilla-mobile/android-components/pull/4936#discussion_r341782953), we wanted to crash on newly added exceptions but we had to list them all out, so we opted for a whitelist instead. Now that we know that we don't need to crash on fatal errors, we're just catching and submitting them and we're at least aware of changes to the system this way.

Are you suggesting we don't crash at all? The exceptions that would go through here would be non-push specific (besides InternalPanic) since we're only logging them with `submitCaughtException`.

Reply:
Let's differentiate between error handling for `onMessageReceived` and for other types of exceptions (one way is to add `onMessageReceivedError` to AutoPushFeature in addition to `onError` - but see last paragraph).

And not crash at all (but rather, log and submit via crash reporter) in case of those types of errors. The other errors we can be more crashy with, I think.

Underlying idea is that it doesn't feel right to me to crash in case of a failure of a fairly isolated system due to an external event.

Also, you'll probably notice that there's currently a bit of a circle dance around error handling - e.g. AbstractFirebasePushService.onMessageReceived -> AutopushFeature.onMessageReceived -> throws -> AbstractFirebasePushService catches -> AutopushFeature.onError.

It seems like this could be simplified. e.g. AutopushFeature.onMessageReceived can just deal with exceptions directly, and if it throws, consumers should re-throw.